### DISCUSSION TITLE
Parent Grain Reconstruction won't work with my data :(

### DISCUSSION THREAD

**Discussion by bitterpea** [mtex-toolbox/mtex#2263] [Ask Anything] [answered]
The Parent grain reconstruction code will not work with my data, I have used the code with the given data and it works fine... It can detect the grain boundaries of the child grains but not the parent ones ... 
I have put in lots of different types of parent-child microstructures in with different orientation relationships, but to no avail :(  I have attached three pictures of the ebsd scans I have been attempting to use, they have all been taken with oxford detectors and are steel welds so i have a variation in the microstructures. Once I get to line 59 it throws out this error too. The file type I am using is .h5oina and my material is Iron bcc. Thanks in advance !!!!!
![image](https://github.com/user-attachments/assets/38c4ebb5-2576-48a6-837e-11d9cfa0c31e)
![image](https://github.com/user-attachments/assets/8b489f88-9fc7-417e-accf-f2ed47929ac9)
![image](https://github.com/user-attachments/assets/f1aee0ce-cc31-4bae-9712-fb7828d6e5ca)
![Screenshot 2024-11-19 105505](https://github.com/user-attachments/assets/213401b7-1164-4b87-8337-4e6fe5155752)

**Comment by zmichels** [comment]
> Once I get to line 59 it throws out this error too.

Although I am not much help on this topic... you might need/want to share the exact code (and maybe the data) you refer to using.

**Comment by nyyssont** [comment]
At least some of the structures look like reconstruction would be possible. The variant graph references the grain map and the error appears when the graph is being generated. This implies there is some sort of issue with the grain map.

It would be helpful if you could let us know which version of MTEX you are using. In addition, posting your full reconstruction script would speed up the process of resolving your issues significantly.

**Comment by bitterpea** [comment]
thank you guys !! I am using version 5.10.2 of mtex. Not quite sure how to share my data as its not the right file type unfortunately 

```matlab
% Load in EBSD data
fileloc = "C:\Users\quee4976\OxfordUniversity\phd\Documents\Aztec data\c4-8\fz\h5oina\fz Specimen 1 Site 1 Map Data 1.h5oina";
ebsd = loadEBSD_h5oina(fileloc);

% grain reconstruction
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'), 'angle', 3*degree, 'minPixel',2);
grains = smooth(grains,5);

% plot the data and the grain boundaries
%changed Iron bcc to Fe (gamma, austenite, bcc)
plot(ebsd('Iron bcc'),ebsd('Iron bcc').orientations,'figSize','large')
hold on
plot(grains.boundary,'linewidth',0.1)
hold off

% set up the job
job = parentGrainReconstructor(ebsd,grains);

% initial guess for the parent to child orientation relationship
job.p2c = orientation.KurdjumovSachs(job.csParent, job.csChild);

close all
histogram(job.calcGBFit./degree,'BinMethod','sqrt')
xlabel('disorientation angle')

job.calcParent2Child

hold on
histogram(job.calcGBFit./degree,'BinMethod','sqrt')
hold off

% compute the misfit for all child to child grain neighbors
[fit,c2cPairs] = job.calcGBFit;

% select grain boundary segments by grain ids
[gB,pairId] = job.grains.boundary.selectByGrainId(c2cPairs);

% plot the child phase
%changed Iron bcc to Fe (gamma, austenite, fcc)
plot(ebsd('Iron bcc'),ebsd('Iron bcc').orientations,'figSize','large','faceAlpha',0.5)

% and on top of it the boundaries colorized by the misfit
hold on;
% scale fit between 0 and 1 - required for |edgeAlpha|
plot(gB, 'edgeAlpha', (fit(pairId) ./ degree - 2.5)./2 ,'linewidth',0.1);
hold off

%this is used for smaller maps
%job.calcVariantGraph('threshold',3.5*degree,'tolerance',2.5*degree) 

%this is used for larger maps 
job.calcVariantGraph('threshold',2.5*degree,'tolerance',2.5*degree,'mergeSimilar')
job.clusterVariantGraph
job.calcVariantGraph('threshold',2.5*degree,'tolerance',2.5*degree)

plot(job.grains,job.votes.prob(:,1))
mtexColorbar

job.calcParentFromVote('minProb',0.5)

% plot the result
plot(job.parentGrains,job.parentGrains.meanOrientation)

%with this you can reconstruct the remaining parents orentations manually
%job.selectInteractive

%this way is to rerun the variant graph approach above a second time but
%with relaxed settings ie with a lower probability 
% compute the votes
job.calcGBVotes('p2c','reconsiderAll','threshold',3*degree,'tolerance',1.5*degree)

% assign parent orientations according to the votes
job.calcParentFromVote

% plot the result
plot(job.parentGrains,job.parentGrains.meanOrientation)

% merge grains with similar orientation
job.mergeSimilar('threshold',7.5*degree);

% plot the result
plot(job.parentGrains,job.parentGrains.meanOrientation)

job.mergeInclusions('maxSize',50);

% plot the result
plot(job.parentGrains,job.parentGrains.meanOrientation)

job.calcVariants

% associate to each packet id a color and plot
color = ind2color(job.transformedGrains.packetId);
plot(job.transformedGrains,color,'faceAlpha',0.5)

hold on
parentGrains = smooth(job.parentGrains,10);
plot(parentGrains.boundary,'linewidth',3)

% outline a specific parent grain
grainSelected = parentGrains(parentGrains.findByLocation([100,80]));

hold on
plot(grainSelected.boundary,'linewidth',3,'lineColor','w')
hold off

% identify childs of the selected parent grain
childGrains = job.grainsPrior(job.mergeId == grainSelected.id);

% plot these childs
plot(childGrains,childGrains.meanOrientation)

% and top the parent grain boundary
hold on
plot(grainSelected.boundary,'linewidth',2)
hold off

% the measured child orientations that belong to parent grain 279
childOri = job.ebsdPrior(childGrains).orientations;

% the orientation of parent grain 279
parentOri = grainSelected.meanOrientation;

% compute variantIds and packeIds
[variantId, packetId] = calcVariantId(parentOri,childOri,job.p2c);

% colorize child orientations by packetId
color = ind2color(packetId);
plotPDF(childOri,color, Miller(0,0,1,childOri.CS),'MarkerSize',2,'all')

% the positions of the parent (001) directions
hold on
plot(parentOri.symmetrise * Miller(0,0,1,parentOri.CS),'markerSize',10,...
  'marker','s','markerFaceColor','w','MarkerEdgeColor','k','linewidth',2)

% the theoretical child variants
childVariants = variants(job.p2c, parentOri);
plotPDF(childVariants, 'markerFaceColor','none','linewidth',1.5,'markerEdgeColor','k')
hold off

parentEBSD = job.calcParentEBSD;

% plot the result
plot(parentEBSD('Iron fcc'),parentEBSD('Iron fcc').orientations,'figSize','large')

% the fit between ebsd child orientation and the reconstructed parent grain
% orientation
plot(parentEBSD, parentEBSD.fit ./ degree,'figSize','large')
mtexColorbar
setColorRange([0,5])
mtexColorMap('LaboTeX')

hold on
plot(job.grains.boundary,'lineWidth',2)
hold off

[parentGrains, parentEBSD.grainId] = calcGrains(parentEBSD('indexed'),'angle',3*degree);

% remove very small grains
parentEBSD(parentGrains(parentGrains.grainSize<10)) = [];

% redo grain reconstruction
[parentGrains, parentEBSD.grainId] = calcGrains(parentEBSD('indexed'),'angle',3*degree);
parentGrains = smooth(parentGrains,10);

plot(ebsd('indexed'),ebsd('indexed').orientations,'figSize','large')

hold on
plot(parentGrains.boundary,'lineWidth',2)
hold off

% fill the holes
F = halfQuadraticFilter;
parentEBSD = smooth(parentEBSD('indexed'),F,'fill',parentGrains);

% plot the parent map
plot(parentEBSD('Iron fcc'),parentEBSD('Iron fcc').orientations,'figSize','large')

% with grain boundaries
hold on
plot(parentGrains.boundary,'lineWidth',2)
hold off
```

**Comment by nyyssont** [comment]
First off, the option "minPixel" does not work in MTEX 5.10.2, since the functionality was only introduced in the latest version.

That being said, can you try the more basic script below:

Please post all error messages fully to get more help.

```
clear
close all

% Load in EBSD data
fileloc = "Your location";
ebsd = loadEBSD_h5oina(fileloc);

% grain reconstruction
[grains,ebsd('indexed').grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
ebsd(grains(grains.grainSize<2)) = 'notIndexed'; % Remove single pixels
ebsd('Iron fcc') = 'notIndexed';
ebsd.grainId = zeros(length(ebsd),1);
%Construct a grain map:
[grains,ebsd('indexed').grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);

% set up the job
job = parentGrainReconstructor(ebsd,grains);

% initial guess for the parent to child orientation relationship
job.p2c = orientation.KurdjumovSachs(job.csParent, job.csChild);

% or optimization
job.calcParent2Child('quantile',0.6)

% variant graph
job.calcVariantGraph('threshold',4*degree,'tolerance',2*degree)
job.clusterVariantGraph
job.calcParentFromVote;
job.mergeSimilar

% check result
figure
plot(job.parentGrains,job.parentGrains.meanOrientation)
```

**Reply by bitterpea** [reply]
cheers! just tried this but when ran as Iron bcc i got these errors (bcc is the only phase i have indexed on the scan) 
![image](https://github.com/user-attachments/assets/c7df45ae-3aae-45be-b0d1-7576cbd0e35b)
and if ran as the og Iron fcc this comes out
![image](https://github.com/user-attachments/assets/77de42cb-f5e7-4ec3-892a-68f43fb0379b)
which is not the full scan (aka the dimensions are wrong) and has not worked :/

**Reply by nyyssont** [reply, marked as answer]
You need to define the parent phase somehow. MTEX tries to guess it from the phase list when you create the parentGrainReconstructor job class object, but if there is no other phase than "Iron bcc" in your dataset then MTEX has nothing to go on with. In this case, you need to add the parent phase manually.

In the future, I would suggest you include "Iron fcc" in the phase list, even you expect that the material is wholly transformed. If you have prepared the specimen well and have set up the microscope and the measurement parameters correctly, the amount of misindexed phases will be very limited. It is good information if there indeed happens to be retained austenite somewhere in the structure. In addition, reconstruction will be easier with the MTEX algorithms.

For now, here is how to add "Iron fcc" to the phase list:

```
clear
close all

% Load in EBSD data
fileloc = "Your location";
ebsd = loadEBSD_h5oina(fileloc);

% Add a phase to the phase list:
ebsd.CSList{end+1} = ...
    crystalSymmetry('m-3m', [3.6599 3.6599 3.6599], ...
  'mineral', 'Iron fcc', 'color', 'red');
ebsd.phaseMap(end+1) = ebsd.phaseMap(end) + 1;

% grain reconstruction
[grains,ebsd('indexed').grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
ebsd(grains(grains.grainSize<2)) = 'notIndexed'; % Remove single pixels
ebsd.grainId = zeros(length(ebsd),1);
%Construct a grain map:
[grains,ebsd('indexed').grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);

% set up the job
job = parentGrainReconstructor(ebsd,grains);

% initial guess for the parent to child orientation relationship
job.p2c = orientation.KurdjumovSachs(job.csParent, job.csChild);

% or optimization
job.calcParent2Child('quantile',0.6)

% variant graph
job.calcVariantGraph('threshold',4*degree,'tolerance',2*degree)
job.clusterVariantGraph
job.calcParentFromVote;
job.mergeSimilar

% check result
figure
plot(job.parentGrains,job.parentGrains.meanOrientation)
```

Please give this code a try and come back with any further error messages.

**Reply by bitterpea** [reply]
omg you legend !!!!!!!!!!!!!!!!!!
for some reason it is flipping it round and i cannot figure out why haha 
i think the angle relationships are a little off but this is sooooo good :) 
![image](https://github.com/user-attachments/assets/a6e92e28-952f-4006-81c2-c4feaa598697)
![image](https://github.com/user-attachments/assets/88d4937a-8c8c-429c-918e-6a79d91f8caa)
![image](https://github.com/user-attachments/assets/c42fbdec-5d38-4bd3-b7d0-f743f9dc9861)
![image](https://github.com/user-attachments/assets/411b21ed-85b4-4279-81da-492d9661f583)
![image](https://github.com/user-attachments/assets/26a8d6ec-d9d0-4bc7-aa2b-67f8aa4a864e)
![image](https://github.com/user-attachments/assets/4511fc2e-911a-4aaf-96a1-331903b657d1)

**Reply by nyyssont** [reply]
Well done. The plotting is controlled by the commands:

```
plotx2east
plotx2west
plotx2north
plotx2south
plotzIntoPlane
plotzOutOfPlane
```

Adding a line with `plotx2east` somewhere before the figure line in the script might result in a plot closer to your expectations. You can try different combinations of plotx2 and plotz commands to see what you like best.

I encourage you to keep trying different things with your reconstructed dataset. Post questions here or make another topic to get further advice.
